### DISCUSSION TITLE
Formatting of document not working

### DISCUSSION THREAD

**Discussion by ydaveluy** [eclipse-langium/langium#1624] [Q&A]
Hello,

I have some formatting issues with my dsl.

Before calling AbstractFormatter.avoidOverlappingEdits I have the following edits:
```
  newText:'', start:'1:1', end:'1:1'
  newText:' ', start:'1:10', end:'1:11'
  newText:'', start:'2:1', end:'2:2'
  newText:'', start:'3:1', end:'3:1'
  newText:'', start:'2:1', end:'2:2'
  newText:'', start:'3:1', end:'3:1'

```

and after:
```
  newText:'', start:'1:1', end:'1:1'
  newText:' ', start:'1:10', end:'1:11'
  newText:'', start:'2:1', end:'2:2'
  newText:'', start:'2:1', end:'2:2'
  newText:'', start:'3:1', end:'3:1'
```

The `newText:'', start:'2:1', end:'2:2'` edit is duplicated which prevent from formatting the whole file.

As a workaround I overrode the avoidOverlappingEdits method:
```ts
    protected override avoidOverlappingEdits(textDocument: TextDocument, textEdits: TextEdit[]): TextEdit[] {
        const edits: TextEdit[] = [];
        for (const edit of textEdits) {
            let last = edits[edits.length - 1];
            while (last) {
                const currentStart = textDocument.offsetAt(edit.range.start);
                const lastEnd = textDocument.offsetAt(last.range.end);
                if (currentStart < lastEnd) {
                    edits.pop();
                    last = edits[edits.length - 1];
                }
                else {
                    break;
                }
            }
            edits.push(edit);
        }
        return edits;
    }
```

Is it an issue in  AbstractFormatter.avoidOverlappingEdits or somewhere else ?

**Comment by msujew** [comment]
@ydaveluy Can you please open an issue for this? While the formatter implementation (i.e. your formatter) shouldn't generate duplicate formattings in the first place (since the formatting can become non-deterministic), the `avoidOverlappingEdits` method should avoid such errors even if they exist.

**Comment by ydaveluy** [comment]
Issues opened.

In my case the duplicated edit is on a hidden ML_COMMENT rule.
This is the text that produce the duplicate formatting
```
catalogue Test           
 /**
*/
```
I don't perform any operation in my formatter on the ML_COMMENT rule.
I will have a look on createHiddenTextEdits to see if creates the duplicates.

**Reply by ydaveluy** [reply]
The createHiddenTextEdits is called twice for the same node:

A firts time here (nextNode is hidden)  : https://github.com/eclipse-langium/langium/blob/b2c40e3a8f06becfa7b18b04cfbef7b7931f164a/packages/langium/src/lsp/formatter.ts#L243 
And a second time here: https://github.com/eclipse-langium/langium/blob/b2c40e3a8f06becfa7b18b04cfbef7b7931f164a/packages/langium/src/lsp/formatter.ts#L253


I locally changed https://github.com/eclipse-langium/langium/blob/b2c40e3a8f06becfa7b18b04cfbef7b7931f164a/packages/langium/src/lsp/formatter.ts#L242 to 
```ts
if (nextNode && !nextNode.hidden)
```
With this change hidden nodes are formatted on the next iteration only
